Source/Scene/ShadowMap.js

```diff
@@ -0,0 +1,397 @@
+                alpha : false
+            },
+            depthMask : true,
+            viewport : new BoundingRectangle(0, 0, this._size, this._size)
+        });
+
+        this.passState = new PassState(scene.context);
+
+        // TODO : determine frustum based on scene
+        var frustum = new OrthographicFrustum();
+        frustum.left = -15.0;
+        frustum.right = 15.0;
+        frustum.bottom = -15.0;
+        frustum.top = 15.0;
+        frustum.near = 1.0;
+        frustum.far = 100.0;
+        this._frustum = frustum;
+
+        // TODO : position camera based on sun direction
+        var camera = new Camera(scene);
+        camera.frustum = frustum;
+        var centerLongitude = -1.31968;
+        var centerLatitude = 0.698874;
+        camera.lookAt(Cartesian3.fromRadians(centerLongitude, centerLatitude), new Cartesian3(0.0, 0.0, 75.0));
+        this.camera = camera;
+
+        this.clearCommand = new ClearCommand({
+            depth : 1.0,
+            renderState : this.renderState,
+            framebuffer : undefined, // Set later
+            owner : this
+        });
+    }
+
+    function destroyTextures(shadowMap) {
+        shadowMap._depthStencilTexture = shadowMap._depthStencilTexture && !shadowMap._depthStencilTexture.isDestroyed() && shadowMap._depthStencilTexture.destroy();
+    }
+
+    function destroyFramebuffers(shadowMap) {
+        shadowMap.framebuffer = shadowMap.framebuffer && !shadowMap.framebuffer.isDestroyed() && shadowMap.framebuffer.destroy();
+    }
+
+    function createTextures(shadowMap, context) {
+        // TODO : Use nearest filtering for testing, not default linear.
+        shadowMap._depthStencilTexture = new Texture({
+            context : context,
+            width : shadowMap._size,
+            height : shadowMap._size,
+            pixelFormat : PixelFormat.DEPTH_STENCIL,
+            pixelDatatype : PixelDatatype.UNSIGNED_INT_24_8
+        });
+    }
+
+    function createFramebuffers(shadowMap, context) {
+        destroyTextures(shadowMap);
+        destroyFramebuffers(shadowMap);
+
+        createTextures(shadowMap, context);
+
+        var framebuffer = new Framebuffer({
+            context : context,
+            depthStencilTexture : shadowMap._depthStencilTexture,
+            destroyAttachments : false
+        });
+
+        shadowMap._framebuffer = framebuffer;
+        shadowMap.passState.framebuffer = framebuffer;
+        shadowMap.clearCommand.framebuffer = framebuffer;
+    }
+
+    function updateFramebuffers(shadowMap, context) {
+        var depthStencilTexture = shadowMap._depthStencilTexture;
+        var textureChanged = !defined(depthStencilTexture) || (depthStencilTexture.width !== shadowMap._size);
+        if (!defined(shadowMap.framebuffer) || textureChanged) {
+            createFramebuffers(shadowMap, context);
+        }
+    }
+
+    ShadowMap.createShadowCastProgram = function(shaderProgram) {
+        // TODO : unused right now
```

Comment:
Also add a `TODO` for handling when a fragment shader does discard or writes an alpha of zero.  For example, see how we generate pick shaders:

https://github.com/AnalyticalGraphicsInc/cesium/blob/master/Source/Renderer/ShaderSource.js#L344-L357

It's also not clear to me yet if we want to move this shader pipeline code there.